processor/batchprocessor/batch_processor.go

```diff
@@ -274,8 +316,77 @@ func (b *shard) resetTimer() {
 func (b *shard) sendItems(trigger trigger) {
 	sent, req := b.batch.splitBatch(b.exportCtx, b.processor.sendBatchMaxSize)
 
-	err := b.batch.export(b.exportCtx, req)
-	if err != nil {
+	var thisBatch []pendingItem
+
+	numItemsBefore := b.totalSent
+	numItemsAfter := b.totalSent + uint64(sent)
+
+	// The current batch can contain items from several different
+	// producers.  Update pending to correctly track contexts
+	// included in the current batch.
+	for len(b.pending) > 0 && numItemsBefore < numItemsAfter {
+		if numItemsBefore+uint64(b.pending[0].numItems) > numItemsAfter {
+			// Waiter only had some items in the current batch
+			partialSent := int(numItemsAfter - numItemsBefore)
+			numItemsBefore = numItemsAfter
+			b.pending[0].numItems -= partialSent
+			thisBatch = append(thisBatch, pendingItem{
+				numItems:  partialSent,
+				parentCtx: b.pending[0].parentCtx,
+			})
+		} else {
+			// This item will be completely processed.
+			numItemsBefore += uint64(b.pending[0].numItems)
+			thisBatch = append(thisBatch, pendingItem{
+				numItems:  b.pending[0].numItems,
+				parentCtx: b.pending[0].parentCtx,
+			})
+
+			// Shift the pending array, to allow it to be re-used.
+			copy(b.pending[0:len(b.pending)-1], b.pending[1:])
+			b.pending[len(b.pending)-1] = pendingItem{}
+			b.pending = b.pending[:len(b.pending)-1]
+		}
+	}
+
+	b.totalSent = numItemsAfter
+
+	var err error
+
+	var parentSpan trace.Span
+	var parent context.Context
+	isSingleCtx := allSameContext(thisBatch)
+
+	// If incoming requests are sufficiently large, there
+	// will be one context, in which case no need to create a new
+	// root span.
+	if isSingleCtx {
+		parent = thisBatch[0].parentCtx
+		parent, parentSpan = b.processor.tracer.Start(parent, "batch_processor/export")
+	} else {
+		spans := parentSpans(thisBatch)
```

Comment:
I think you are suggesting a simplification that would break the current invariants. Today, this processor will only form batches less than the batch size when the timeout triggers. You are suggesting to flush a small batch when adding new data would cause it to exceed the max batch size, I think.